HW1/P6/P6.py

```diff
@@ -0,0 +1,87 @@
+import findspark
+findspark.init()
+import pyspark
+import re
+
+sc = pyspark.SparkContext(appName="P6")
+sc.setLogLevel('ERROR')
+
+is_invalid = r"(?:\A[0-9]+\Z)|(?:\A[A-Z]+\Z)|(?:\A[A-Z]+\.\Z)"
+
+#read the file, and split based on spaces
+shakespeare = sc.textFile("Shakespeare.txt")
+words = shakespeare.flatMap(lambda x: x.split())
+
+#filter out words which are invalid based on our regex
+words = words.filter(lambda x: not (re.match(is_invalid, x) != None)).collect()
+
+#convert the words into a set of 3tuples
+tuples = sc.parallelize([((words[i], words[i+1], words[i+2]), 1) for i in xrange(len(words)-2)])
+
+def combine(count1, count2):
+	return count1+count2
+
+tuples_counts = tuples.reduceByKey(combine)
+
+tuples_counts = tuples_counts.map(lambda ((w1,w2,w3),count): ((w1,w2),[(w3,count)]))
+
+#tuples_counts_list = tuples_counts.groupByKey()
+tuples_counts_list = tuples_counts.reduceByKey(lambda list1, list2: list1+list2)
+
+nowis = tuples_counts_list.filter(lambda ((w1,w2),v): w1=="Now" and w2=="is").collect()
+
+#for the Spark 1.5.0 bug
+tuples_counts_list = tuples_counts_list.map(lambda x:x).cache()
+
+import numpy as np
+
+#words_weights is a list of (word, weight) tuples
+def sample_randomly(words_weights):	
+	total_word_count = float(sum([x[1] for x in words_weights]))
+	probabilities = [x[1]/total_word_count for x in words_weights]
+
+	#from http://stackoverflow.com/questions/6432499/how-to-do-weighted-random-sample-of-categories-in-python
+	word_index = np.array(probabilities).cumsum().searchsorted(np.random.rand())
```

Comment:
Another way to do random samples in this case is to repeat values that have a higher count (eg (is,3) -> [is,is,is]), then sample out of the larger list. Since these values tend to be rather small in this case, I do not think it is a big space issue, but would become one if the weights are much larger than the usual 1.